pkg/incentives/events.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package incentives
+
+import (
+	"context"
+	"sync"
+)
+
+type phaseType int
+
+const (
+	commit phaseType = iota
+	reveal
+	claim
+	sample
+	sampleEnd
+)
+
+func (p phaseType) string() string {
+	switch p {
+	case commit:
+		return "commit"
+	case reveal:
+		return "reveal"
+	case claim:
+		return "claim"
+	case sample:
+		return "sample"
+	case sampleEnd:
+		return "sampleEnd"
+	default:
+		return "unknown"
+	}
+}
+
+type events struct {
+	on       map[phaseType][]func(context.Context)
```

Comment:
Not sure why the three separate maps are needed. If you create `event` struct with three fields, that will do:
```golang
type event struct {
	on         []func(context.Context)
	ctx        context.Context
	cancelFunc context.CancelFunc
}
```
And then simply `map[phaseType]*event`.

Reply:
good suggestion <3